failpoint.go

```diff
@@ -35,59 +35,76 @@ type (
 	// failpoint will not to be evaluated.
 	Hook func(ctx context.Context, fpname string) bool
 
-	failpoint struct {
+	// Failpoint TODO
+	Failpoint struct {
 		mu       sync.RWMutex
 		t        *terms
 		waitChan chan struct{}
 	}
 )
 
 // Pause will pause until the failpoint is disabled.
-func (fp *failpoint) Pause() {
+func (fp *Failpoint) Pause() {
 	<-fp.waitChan
 }
 
-// WithHook binds a hook to a new context which is based on the `ctx` parameter
-func WithHook(ctx context.Context, hook Hook) context.Context {
-	return context.WithValue(ctx, failpointCtxKey, hook)
+// Enable sets a failpoint to a given failpoint description.
+func (fp *Failpoint) Enable(inTerms string) error {
+	t, err := newTerms(inTerms, fp)
+	if err != nil {
+		return err
+	}
+	fp.mu.Lock()
+	fp.t = t
+	fp.waitChan = make(chan struct{})
+	fp.mu.Unlock()
+	return nil
 }
 
-// EvalContext evaluates a failpoint's value, and calls hook if the context is
-// not nil and contains hook function. It will return the evaluated value and
-// true if the failpoint is active. Always returns false if ctx is nil or context
-// does not contains hook function
-func EvalContext(ctx context.Context, fpname string) (Value, bool) {
-	if ctx == nil {
-		return nil, false
+// EnableAndLock enables and locks the failpoint, the lock prevents
+// the failpoint to be evaled. It returns an unlock function or an
+// error if there is. It is useful when enables a panic failpoint
+// and does some post actions before the failpoint being evaled.
+func (fp *Failpoint) EnableAndLock(inTerms string) (func(), error) {
```

Comment:
Do we need to export the `EnableAndLock` method? It will be a bad idea to let users use lock externally.

Reply:
Yes, I think so. The `failpoint` package could be used as a lower level for other implementations, it is necessary to use this function just like we use it.

Reply:
I agree that this function looks too dangerous for misuse. What if we change the signature to

```go
func (fp *Failpoint) EnableWith(inTerms string, actionsWhileLocked func()) error
```

?

Reply:
We are supplying a library, not a service, so I don't think it is a matter of danger. Otherwise, I like your signature, it indeed reduces the risk of misuse.